File: src/cpp_action/src/count_until_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/count_until.hpp"

using CountUntil = my_robot_interfaces::action::CountUntil;
using CountUntilGoalHandle = rclcpp_action::ServerGoalHandle<CountUntil>;
using namespace std::placeholders;

class CountUntilServerNode : public rclcpp::Node 
{
public:
    CountUntilServerNode() : Node("count_until_server") 
    {
        count_until_server_ = rclcpp_action::create_server<CountUntil>(
            this,
            "count_until",
            std::bind(&CountUntilServerNode::goal_callback, this, _1, _2),
            std::bind(&CountUntilServerNode::cancel_callback, this, _1),
            std::bind(&CountUntilServerNode::handle_accepted_callback_callback, this, _1)
        );
        RCLCPP_INFO(get_logger(), "Action Server has been started");
    }
 
private:
    rclcpp_action::Server<CountUntil>::SharedPtr count_until_server_;

    rclcpp_action::GoalResponse goal_callback(
        const rclcpp_action::GoalUUID &uuid, 
        std::shared_ptr<const CountUntil::Goal> goal){
            (void) uuid;
            RCLCPP_INFO(get_logger(), "Received a goal");
            if(goal->target_number <= 0.0){
                RCLCPP_WARN(get_logger(), "target number <= 0 \nRejecting the goal");
                return rclcpp_action::GoalResponse::REJECT;
            }
            RCLCPP_INFO(get_logger(), "Accepting the goal");
            return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
        }

    rclcpp_action::CancelResponse cancel_callback(
        const std::shared_ptr<CountUntilGoalHandle> goal_handle){
            (void) goal_handle;
            return rclcpp_action::CancelResponse::ACCEPT;
        }
    
    void handle_accepted_callback_callback(
        const std::shared_ptr<CountUntilGoalHandle> goal_handle){
            RCLCPP_INFO(get_logger(), "Executing the goal");
            execute_goal(goal_handle);
        }

    void execute_goal(
        const std::shared_ptr<CountUntilGoalHandle> goal_handle){
            // Get request from goal
            int target_number = goal_handle->get_goal()->target_number;
            double period = goal_handle->get_goal()->period;

            //Execute the goal
            int counter=0;
            auto feedback = std::make_shared<CountUntil::Feedback>();
            rclcpp::Rate loop_rate(1.0/period);
            for(int i=0; i<target_number; ++i){
                counter++;
                RCLCPP_INFO(get_logger(), "Current number is : %d", counter);
                feedback->current_number = counter;
                goal_handle->publish_feedback(feedback);
                loop_rate.sleep();
            }

            //Set final state and return result
            auto result = std::make_shared<CountUntil::Result>();
            result->reached_number = counter;
            goal_handle->succeed(result);
    }
};
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CountUntilServerNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/py_action/py_action/count_until_server.py
#!/usr/bin/env python3
import rclpy
import time
import threading
from rclpy.node import Node
from rclpy.action import ActionServer, GoalResponse, CancelResponse
from rclpy.action.server import ServerGoalHandle
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup
from my_robot_interfaces.action import CountUntil
 
 
class CountUntilServerNode(Node): 
    def __init__(self):
        super().__init__("count_until_server")
        self.goal_handle_: ServerGoalHandle = None
        self.goal_lock_ = threading.Lock()
        self.count_until_server_ =  ActionServer(self, CountUntil, "count_until", goal_callback=self.goal_callback, cancel_callback=self.cancelGoal_callback, execute_callback=self.execute_callback, callback_group=ReentrantCallbackGroup())
        self.get_logger().info("Action Server has been started")

    # Used for validating the arguments of the request
    def goal_callback(self, goal_request: CountUntil.Goal):
        self.get_logger().info("Received a Goal")

        # Policy: Refuse new goal if current goal is active
        #with self.goal_lock_:
        #    if self.goal_handle_ is not None and self.goal_handle_.is_active:
        #        self.get_logger().warn("A goal is already active, rejecting new gaol")
        #        return GoalResponse.REJECT

        # Validate arguments of the call
        if goal_request.target_number <= 0:
            self.get_logger().warn("Rejecting the Goal")
            return GoalResponse.REJECT
        self.get_logger().info("Accepting the Goal")
        return GoalResponse.ACCEPT


    def execute_callback(self, goal_handle: ServerGoalHandle):
        with self.goal_lock_:
            self.goal_handle_ = goal_handle

        # Get request from the goal
        target_number = goal_handle.request.target_number
        period = goal_handle.request.period

        # Execute the action
        self.get_logger().info("Executing the goal")
        feedback = CountUntil.Feedback()
        result = CountUntil.Result()
        counter = 0
        for i in range(target_number):
            if goal_handle.is_cancel_requested:
                self.get_logger().info("Cancelling the goal")
                goal_handle.canceled()
                result.reached_number = counter
                return result
            counter+=1
            self.get_logger().info(str(counter))
            feedback.current_number = counter
            goal_handle.publish_feedback(feedback)
            time.sleep(period)
            
        
        # Once done, set goal final state
        goal_handle.succeed()

        # Send the result
        result.reached_number = counter
        return result
    
    def cancelGoal_callback(self, goal_handle: ServerGoalHandle):
        self.get_logger().info("Received a cancelRequest")
        return CancelResponse.ACCEPT # Or reject as required
 
def main(args=None):
    rclpy.init(args=args)
    node = CountUntilServerNode() 
    rclpy.spin(node, MultiThreadedExecutor())
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()